clientcmd-based kubeconfig handling for single-context functions

skitt opened the pull request:
See individual commits for details.

This relates to https://github.com/submariner-io/enhancements/issues/79

skitt committed:
Split up RestConfigProducer instances

This ensures each command has its own producer setup, and will help
with the kubeconfig handling rewrite.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Wrap kubeconfig flag handling in restconfig.Producer

Moving the execution function into restconfig.Producer will ensure
that flags can be handled entirely inside this package, reducing the
exposure of internals to subctl commands, and the knowledge required
to implement commands.

For now only single-context commands are handled, multi-context
commands will come later. configAndOverrides will be useful then.

This also removes ForClusters() and Clients() which are effectively
unused, and unexports ClusterNameFromContext() which is only used in
the restconfig package.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Use wrapped kubeconfig flag handling for export

This provides support for contexts in export service.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Use wrapped kubeconfig flag handling for join

restconfig.GetClusterID() is no longer used, remove it.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Use wrapped kubeconfig flag handling for uninstall

The namespace flag is handled by restconfig.Producer.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Use wrapped kubeconfig flag handling for deploy-broker

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Use wrapped kubeconfig flag handling for cloud-prepare

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Use wrapped kubeconfig flag handling for unexport

The namespace flag is handled by restconfig.Producer.

This removes the last use of restconfig.AddKubeContextFlag(), so that
is removed too. restconfig.ClientConfig() is no longer used outside
restconfig, so this unexports that.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt commented on internal/restconfig/restconfig.go:
```diff
@@ -26,58 +26,134 @@ import (
+	}
+}
+
+type PerContextFn func(clusterName, namespace string, clientProducer client.Producer) error
```
I’ll rework this to use `cluster.Info`.

tpantelis replied:
I assume in another PR?

skitt replied:
No, in this PR.

skitt committed:
Move internal/cluster to pkg/cluster

This avoids having two cluster packages (in internal and pkg). Since
pkg/cluster is used by other public functions, we need to keep it
public.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt commented:
This works locally, I’m looking into why it fails in CI.

tpantelis commented on pkg/diagnose/connections.go:
```diff
@@ -20,7 +20,7 @@ package diagnose
 import (
 	"github.com/submariner-io/admiral/pkg/reporter"
-	"github.com/submariner-io/subctl/pkg/cluster"
+	"github.com/submariner-io/subctl/internal/cluster"
```
`cluster.Info` is referenced in the public `Connections` function (and other public functions) so if it's moved internal I think it would be problematic for users outside the repo. Perhaps we can move the _internal/cluster_  functions to _pkg/cluster_ instead.

skitt committed:
cluster.NewInfo: don't fail if CRD is missing

Retrieving Submariner and/or ServiceDiscovery can fail because the CRD
itself hasn't been installed; this is a valid situation and shouldn't
result in an error.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt commented on internal/restconfig/restconfig.go:
```diff
@@ -26,58 +26,140 @@ import (
+// Runs the given function on the selected context.
+// The status reporter is only used for errors specific to this function;
+// the per-context function is expected to perform its own status reporting.
+func (rcp *Producer) RunOnSelectedContext(function PerContextFn, status reporter.Interface) error {
```
Since this ends up tied to a status reporter anyway, I wonder if I shouldn’t just add `status` to `PerContextFn`.

tpantelis replied:
That seems reasonable.

tpantelis commented on pkg/cluster/cluster_test.go:
```diff
@@ -23,7 +23,7 @@ import (
 
 	. "github.com/onsi/ginkgo"
```
Maybe a nit but since these functions as related to cluster IDs perhaps we could name the files accordingly, ie _id.go_ and _id_test.go_.

mkolesnik reviewed (commented):
@skitt do you reckon this PR needs the release note, or do you plan to add it on a later PR?

skitt committed:
Add release notes for single-context kubeconfig

Signed-off-by: Stephen Kitt <skitt@redhat.com>

mkolesnik commented on internal/restconfig/restconfig.go:
```diff
@@ -52,24 +54,112 @@ type RestConfig struct {
+		return function(clusterInfo, "default", status)
+	}
+
+	if rcp.defaultClientConfig != nil {
```
Shouldn't it error out if the config is not properly initialized?

I feel like we might miss a bug otherwise, as it'll just return no error but also not do anything AFAICT

skitt replied:
Ah yes, I meant to address that earlier and forgot, thanks for noticing!

mkolesnik replied:
Thanks, it's usually better (i.e. makes the code more maintainable) to address the exceptional case in the if, not the normal flow, so I'd expect it to look something like this:
```suggestion
	// If we get here, no context was set up, which means SetupFlags() wasn't called
	if rcp.defaultClientConfig == nil {
		return status.Error(errors.New("no context provided (this is a programming error)"), "")
	}
```

mkolesnik commented on internal/restconfig/restconfig.go:
```diff
@@ -66,6 +66,7 @@ type Producer struct {
 	defaultClientConfig *configAndOverrides
 	inCluster           bool
 	namespaceFlag       bool
+	defaultNamespace    *string
```
nit: Saving as pointer here is a kind of unnecessary optimization IMHO

skitt replied:
It’s not an optimisation, it’s so that we can determine whether a value was given or not (_i.e._ `nil` is significant).

mkolesnik replied:
Ah OK I found it, makes sense

mkolesnik commented on cmd/subctl/deploybroker.go:
```diff
@@ -40,31 +42,29 @@ var (
-			stringset.New(deployflags.BrokerSpec.Components...), deployflags.BrokerSpec.DefaultCustomDomains, status)
-		exit.OnError(err)
+		exit.OnError(deployRestConfigProducer.RunOnSelectedContext(
+			func(clusterInfo *cluster.Info, namespace string, status reporter.Interface) error {
+				if err := deploy.Broker(&deployflags, clusterInfo.ClientProducer, status); err != nil {
+					return err // nolint:wrapcheck // No need to wrap errors here.
+				}
+
+				return broker.WriteInfoToFile( // nolint:wrapcheck // No need to wrap errors here.
+					clusterInfo.RestConfig, deployflags.BrokerNamespace, ipsecSubmFile,
+					stringset.New(deployflags.BrokerSpec.Components...), deployflags.BrokerSpec.DefaultCustomDomains, status)
+			}, cli.NewReporter()))
```
I find it relatively hard to read this anonymous function with such a high level of indentation.
Do you think we can extract it to a named function, for easier code maintainability?

mkolesnik commented on cmd/subctl/join.go:
```diff
@@ -55,34 +58,31 @@ var joinCmd = &cobra.Command{
-		determineServiceCIDR(networkDetails, status)
+				if brokerInfo.IsConnectivityEnabled() && labelGateway {
+					possiblyLabelGateway(clusterInfo.ClientProducer.ForKubernetes(), status)
+				}
 
-		if brokerInfo.IsConnectivityEnabled() && labelGateway {
-			possiblyLabelGateway(clientProducer.ForKubernetes(), status)
-		}
-
-		if joinFlags.CustomDomains == nil && brokerInfo.CustomDomains != nil {
-			joinFlags.CustomDomains = *brokerInfo.CustomDomains
-		}
+				if joinFlags.CustomDomains == nil && brokerInfo.CustomDomains != nil {
+					joinFlags.CustomDomains = *brokerInfo.CustomDomains
+				}
 
-		err = join.ClusterToBroker(brokerInfo, &joinFlags, clientProducer, status)
-		exit.OnError(err)
+				return join.ClusterToBroker( // nolint:wrapcheck // No need to wrap errors here.
+					brokerInfo, &joinFlags, clusterInfo.ClientProducer, status)
```
I find it relatively hard to read this anonymous function with such a high level of indentation.
Do you think we can extract it to a named function, for easier code maintainability?

mkolesnik replied:
Hmm too bad the broker still needs to be accessed separately. complicating the extraction somewhat. but it's still less unruly :)

mkolesnik replied:
Theoretically you could get an ellipsis or args as optional last argument to the `RunOnSelectedContext` function and pass it on to the nested function, but it might only serve this one use case, while complicating the function signature for everyone :thinking:

skitt replied:
Passing `args` would also be useful in `export` and `unexport`, but I’m not sure it’s worth it.

mkolesnik commented on cmd/subctl/uninstall.go:
```diff
@@ -23,55 +23,52 @@ import (
-
-		exit.OnError(uninstall.All(clientProducer, config.ClusterName, uninstallOptions.namespace, status))
+		exit.OnError(uninstallRestConfigProducer.RunOnSelectedContext(
+			func(clusterInfo *cluster.Info, namespace string, status reporter.Interface) error {
+				if !uninstallOptions.noPrompt {
+					result := false
+					prompt := &survey.Confirm{
+						Message: fmt.Sprintf(
+							"This will completely uninstall Submariner from the cluster %q. Are you sure you want to continue?",
+							clusterInfo.Name),
+					}
+
+					_ = survey.AskOne(prompt, &result)
+
+					if !result {
+						return nil
+					}
+				}
+
+				return uninstall.All( // nolint:wrapcheck // No need to wrap errors here.
+					clusterInfo.ClientProducer, clusterInfo.Name, namespace, status)
```
I find it relatively hard to read this anonymous function with such a high level of indentation.
Do you think we can extract it to a named function, for easier code maintainability?

mkolesnik reviewed (commented):
Thanks for this :)

I had a few comments/questions

mkolesnik commented on cmd/subctl/unexport.go:
```diff
@@ -41,35 +43,24 @@ var (
+			exit.OnError(unexportRestConfigProducer.RunOnSelectedContext(
+				func(clusterInfo *cluster.Info, namespace string, status reporter.Interface) error {
+					mcsClient, err := mcsclient.NewForConfig(clusterInfo.RestConfig)
+					exit.OnError(status.Error(err, "Error creating client"))
```
Do we need `exit.OnError` here?

In other places you removed it

mkolesnik commented on internal/restconfig/restconfig.go:
```diff
@@ -316,36 +389,34 @@ func (rcp *Producer) ClientConfig() clientcmd.ClientConfig {
+		}
 
-	exit.OnErrorWithMessage(err, fmt.Sprintf("Error retrieving Submariner object %s", names.SubmarinerCrName))
+		exit.OnErrorWithMessage(err, fmt.Sprintf("Error retrieving Submariner object %s", names.SubmarinerCrName))
```
Is this left here intentionally or should it return the error?

skitt replied:
It’s left intentionally for now, this function is used for old flag-handling and new flag-handling.

mkolesnik commented on cmd/subctl/azure.go:
```diff
@@ -16,28 +16,36 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// nolint:dupl // Similar code in aws.go, azure.go, rhos.go, but not duplicate
```
Just a question, why not do the flag init once for all `cloud` commands? They all need the same set of flags.

skitt replied:
Fair question, I’ve updated the PR to share the setup. Thanks!

skitt committed:
Merge branch 'devel' into kubeconfig-single-context
